http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/DataprocInterpreter.scala

```diff
@@ -376,40 +404,51 @@ class DataprocInterpreter[F[_]: Timer: Async: Parallel: ContextShift: Logger](
+  def setupDataprocImageGoogleGroup(implicit ev: Ask[F, AppContext]): F[Unit] =
+    createDataprocImageUserGoogleGroupIfItDoesntExist >>
       addIamRoleToDataprocImageGroup
 
   /**
    * Add the user's service account to the Google group.
    * This group has compute.imageUser role on the custom Dataproc image project,
    * which allows the user's cluster to pull the image.
    */
-  def updateDataprocImageGroupMembership(googleProject: GoogleProject, createCluster: Boolean): F[Unit] =
+  def updateDataprocImageGroupMembership(googleProject: GoogleProject,
+                                         createCluster: Boolean)(implicit ev: Ask[F, AppContext]): F[Unit] =
     parseImageProject(config.dataprocConfig.customDataprocImage).traverse_ { imageProject =>
       for {
         count <- inTransaction(clusterQuery.countActiveByProject(googleProject))
         // Note: Don't remove the account if there are existing active clusters in the same project,
         // because it could potentially break other clusters. We only check this for the 'remove' case.
         _ <- if (count > 0 && !createCluster) {
-          Async[F].unit
+          F.unit
         } else {
           for {
-            projectNumberOptIO <- Async[F].liftIO(
+            projectNumberOptIO <- F.liftIO(
               IO.fromFuture(IO(googleProjectDAO.getProjectNumber(googleProject.value)))
             )
-            projectNumber <- Async[F].liftIO(
+            projectNumber <- F.liftIO(
               IO.fromEither(projectNumberOptIO.toRight(ClusterIamSetupException(imageProject)))
             )
             // Note that the Dataproc service account is used to retrieve the image, and not the user's
             // pet service account. There is one Dataproc service account per Google project. For more details:
             // https://cloud.google.com/dataproc/docs/concepts/iam/iam#service_accounts
+
+            // Note we add both service-[project-number]@dataproc-accounts.iam.gserviceaccount.com and
+            // [project-number]@cloudservices.gserviceaccount.com to the group because both seem to be
+            // used in different circumstances (the latter seems to be used for adding preemptibles, for example).
             dataprocServiceAccountEmail = WorkbenchEmail(
               s"service-${projectNumber}@dataproc-accounts.iam.gserviceaccount.com"
             )
             _ <- updateGroupMembership(config.groupsConfig.dataprocImageProjectGroupEmail,
                                        dataprocServiceAccountEmail,
                                        createCluster)
+            apiServiceAccountEmail = WorkbenchEmail(
+              s"${projectNumber}@cloudservices.gserviceaccount.com"
+            )
+            _ <- updateGroupMembership(config.groupsConfig.dataprocImageProjectGroupEmail,
+                                       apiServiceAccountEmail,
+                                       createCluster)
```

Comment:
This is the fix to the preemptible creation issue, correct?

Reply:
Yup. I have no idea why preemptibles use a different SA than workers to pull the Dataproc image.

Reply:
🤯